src/v/raft/id_allocator_stm.cc

```diff
@@ -0,0 +1,249 @@
+// Copyright 2020 Vectorized, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "raft/id_allocator_stm.h"
+
+#include "raft/errc.h"
+#include "raft/logger.h"
+#include "raft/types.h"
+#include "storage/record_batch_builder.h"
+
+#include <seastar/core/future.hh>
+
+namespace raft {
+
+template<typename T>
+model::record_batch serialize_cmd(T t, model::record_batch_type type) {
+    storage::record_batch_builder b(type, model::offset(0));
+    iobuf key_buf;
+    reflection::adl<uint8_t>{}.to(key_buf, T::record_key);
+    iobuf v_buf;
+    reflection::adl<T>{}.to(v_buf, std::move(t));
+    b.add_raw_kv(std::move(key_buf), std::move(v_buf));
+    return std::move(b).build();
+}
+
+id_allocator_stm::id_allocator_stm(
+  ss::logger& logger, raft::consensus* c, config::configuration& config)
+  : raft::state_machine(c, logger, ss::default_priority_class())
+  , _config(config)
+  , _last_seq_tick(0)
+  , _c(c) {}
+
+ss::future<id_allocator_stm::stm_allocation_result>
+id_allocator_stm::allocate_id_and_wait(
+  model::timeout_clock::time_point timeout) {
+    auto prelude = ss::now();
+    auto range = _config.id_allocator_batch_size.value();
+
+    if (_last_allocated_range > 0) {
+        auto allocated_id = _last_allocated_base;
+        _last_allocated_range -= 1;
+        _last_allocated_base += 1;
+
+        return ss::make_ready_future<stm_allocation_result>(
+          stm_allocation_result{allocated_id, raft::errc::success});
+    }
+
+    if (_processed > _config.id_allocator_log_capacity.value()) {
+        sequence_id seq = sequence_id{
+          _run_id.value(), _c->self(), ++_last_seq_tick};
+        prelude = replicate_and_wait(prepare_truncation_cmd{seq}, timeout, seq)
+                    .then([this](bool replicated) {
+                        if (replicated) {
+                            (void)ss::with_gate(_gate, [this] {
+                                return replicate(serialize_cmd(
+                                  execute_truncation_cmd{
+                                    _prepare_offset, _prepare_state},
+                                  id_allocator_stm_batch_type));
+                            });
+                        }
+                    });
+    } else if (_should_cache) {
+        (void)ss::with_gate(_gate, [this] {
+            return replicate(serialize_cmd(
+              execute_truncation_cmd{_prepare_offset, _prepare_state},
+              id_allocator_stm_batch_type));
+        });
```

Comment:
for these gates. are we sure we don't need to handle gate_closed exceptions.

Reply:
Good catch!